In [1]:
import pandas as pd
import numpy as np
import requests
from bs4 import BeautifulSoup
import matplotlib.pyplot as plt
import missingno as msno


In [2]:
fake = pd.read_csv('fake.csv')
true = pd.read_csv('true.csv')

In [3]:
fake.shape

(23481, 4)

In [4]:
true.shape

(21417, 4)

Using the shapes I see that the true stories have 21417 rows and the fake has 23481 rows.  Using this I am going to use 70% of each to train the data and merge the remaining 30% into a new dataframe two see if the program can differentiate the two.  I went ahead and used a calculator, but will try to do this in here.

In [5]:
test =pd.DataFrame() 

In [6]:
def dovide(x,y):
    n=70
    m=30
    x = x.head(int(len(x)*(n/100)))
    y = y.head(int(len(y)*(n/100)))
    fir =y.tail(int(len(y)*(m/100)))
    nex = x.tail(int(len(x)*(m/100)))
    test = pd.concat([fir, nex], ignore_index=True)
    return test
    
    

In [7]:
test = dovide(fake,true)

In [8]:
test.shape

(9427, 4)

In [9]:
fake = fake.head(int(len(fake)*(70/100)))
true = true.head(int(len(true)*(70/100)))

In [10]:
print(test.shape)
print(true.shape)
print(fake.shape)

(9427, 4)
(14991, 4)
(16436, 4)


there was some data loss there, but it was 4000 out of the 40000 total rows, I will continue for now and discuss why later.

In [11]:
print('test'+test.head())
print('fake'+fake.head())
print('true'+true.head())

                                               title  \
0  testSupreme Court temporarily blocks Louisiana...   
1  testWhite House adviser Rhodes to meet with Cu...   
2  testForeign policy thinkers praised by Trump h...   
3  testRepublican candidate Trump reverses stance...   
4  testLatest Hillary Clinton bio for kids highli...   

                                                text           subject  \
0  testWASHINGTON (Reuters) - The Supreme Court, ...  testpoliticsNews   
1  testWASHINGTON (Reuters) - President Barack Ob...  testpoliticsNews   
2  testWASHINGTON (Reuters) - If Americans were l...  testpoliticsNews   
3  testWASHINGTON (Reuters) - Republican presiden...  testpoliticsNews   
4  testPOTOMAC, Md. (Reuters) - As Hillary Clinto...  testpoliticsNews   

                 date  
0  testMarch 4, 2016   
1  testMarch 4, 2016   
2  testMarch 4, 2016   
3  testMarch 4, 2016   
4  testMarch 4, 2016   
                                               title  \
0  fake Donald Tru

In [12]:
test.index

RangeIndex(start=0, stop=9427, step=1)

In [13]:
test=test.reset_index()

In [14]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9427 entries, 0 to 9426
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   index    9427 non-null   int64 
 1   title    9427 non-null   object
 2   text     9427 non-null   object
 3   subject  9427 non-null   object
 4   date     9427 non-null   object
dtypes: int64(1), object(4)
memory usage: 368.4+ KB


In [15]:
ptest=pd.pivot_table(test,values='text',index=['date','title'], columns = ['subject'],aggfunc='first')

In [16]:
ptest.head()

subject                                                                                           Government News  \
date        title                                                                                                   
Apr 1, 2015 MUSLIM WOMAN ARRESTED FOR SPITTING ON HER FELLO...                                                NaN   
Apr 1, 2016 LAWSUIT SEEKS OBAMA Info In Corruption Case: “T...                                                NaN   
            WHILE WE’RE DEBATING CRUZ VS TRUMP: Listen To W...                                                NaN   
Apr 1, 2017 BONKERS BERNIE SANDERS: Prioritizing Jobs Over ...  https://www.youtube.com/watch?v=GPqQIlWksbgVer...   
            BUSTED! CLINTON ADVISOR JOHN PODESTA Paid BIG M...  Russia gave John Podesta 35 millions dollars (...   

subject                                                                                                  politics  \
date        title                                                                                                   
Apr 1, 2015 MUSLIM WOMAN ARRESTED FOR SPITTING ON HER FELLO...  This woman s having trouble entering the Walma...   
Apr 1, 2016 LAWSUIT SEEKS OBAMA Info In Corruption Case: “T...  Thank goodness for Judicial Watch! Over and ov...   
            WHILE WE’RE DEBATING CRUZ VS TRUMP: Listen To W...  AUDIO OF KARL ROVE ON HUGH HEWITT   Picking   ...   
Apr 1, 2017 BONKERS BERNIE SANDERS: Prioritizing Jobs Over ...                                                NaN   
            BUSTED! CLINTON ADVISOR JOHN PODESTA Paid BIG M...                                                NaN   

subject                                                        politicsNews  \
date        title                                                             
Apr 1, 2015 MUSLIM WOMAN ARRESTED FOR SPITTING ON HER FELLO...          NaN   
Apr 1, 2016 LAWSUIT SEEKS OBAMA Info In Corruption Case: “T...          NaN   
            WHILE WE’RE DEBATING CRUZ VS TRUMP: Listen To W...          NaN   
Apr 1, 2017 BONKERS BERNIE SANDERS: Prioritizing Jobs Over ...          NaN   
            BUSTED! CLINTON ADVISOR JOHN PODESTA Paid BIG M...          NaN   

subject                                                        worldnews  
date        title                                                         
Apr 1, 2015 MUSLIM WOMAN ARRESTED FOR SPITTING ON HER FELLO...       NaN  
Apr 1, 2016 LAWSUIT SEEKS OBAMA Info In Corruption Case: “T...       NaN  
            WHILE WE’RE DEBATING CRUZ VS TRUMP: Listen To W...       NaN  
Apr 1, 2017 BONKERS BERNIE SANDERS: Prioritizing Jobs Over ...       NaN  
            BUSTED! CLINTON ADVISOR JOHN PODESTA Paid BIG M...       NaN

In [17]:
ptest.tail()

subject                                                                                                                                  Government News  \
date                                               title                                                                                                   
Sep 9, 2017                                        What Sign Language is This? Man Causes A Stir O...  The man giving out sign language during Govern...   
https://100percentfedup.com/12-yr-old-black-con... https://100percentfedup.com/12-yr-old-black-con...                                                NaN   
https://100percentfedup.com/video-hillary-asked... https://100percentfedup.com/video-hillary-asked...                                                NaN   
https://fedup.wpengine.com/wp-content/uploads/2... https://fedup.wpengine.com/wp-content/uploads/2...                                                NaN   
https://fedup.wpengine.com/wp-content/uploads/2... https://fedup.wpengine.com/wp-content/uploads/2...                                                NaN   

subject                                                                                                                                         politics  \
date                                               title                                                                                                   
Sep 9, 2017                                        What Sign Language is This? Man Causes A Stir O...                                                NaN   
https://100percentfedup.com/12-yr-old-black-con... https://100percentfedup.com/12-yr-old-black-con...  https://100percentfedup.com/12-yr-old-black-co...   
https://100percentfedup.com/video-hillary-asked... https://100percentfedup.com/video-hillary-asked...  https://100percentfedup.com/video-hillary-aske...   
https://fedup.wpengine.com/wp-content/uploads/2... https://fedup.wpengine.com/wp-content/uploads/2...  https://fedup.wpengine.com/wp-content/uploads/...   
https://fedup.wpengine.com/wp-content/uploads/2... https://fedup.wpengine.com/wp-content/uploads/2...  https://fedup.wpengine.com/wp-content/uploads/...   

subject                                                                                               politicsNews  \
date                                               title                                                             
Sep 9, 2017                                        What Sign Language is This? Man Causes A Stir O...          NaN   
https://100percentfedup.com/12-yr-old-black-con... https://100percentfedup.com/12-yr-old-black-con...          NaN   
https://100percentfedup.com/video-hillary-asked... https://100percentfedup.com/video-hillary-asked...          NaN   
https://fedup.wpengine.com/wp-content/uploads/2... https://fedup.wpengine.com/wp-content/uploads/2...          NaN   
https://fedup.wpengine.com/wp-content/uploads/2... https://fedup.wpengine.com/wp-content/uploads/2...          NaN   

subject                                                                                               worldnews  
date                                               title                                                         
Sep 9, 2017                                        What Sign Language is This? Man Causes A Stir O...       NaN  
https://100percentfedup.com/12-yr-old-black-con... https://100percentfedup.com/12-yr-old-black-con...       NaN  
https://100percentfedup.com/video-hillary-asked... https://100percentfedup.com/video-hillary-asked...       NaN  
https://fedup.wpengine.com/wp-content/uploads/2... https://fedup.wpengine.com/wp-content/uploads/2...       NaN  
https://fedup.wpengine.com/wp-content/uploads/2... https://fedup.wpengine.com/wp-content/uploads/2...       NaN

In [18]:
ptest = ptest.drop(ptest.index[len(ptest)-1])
ptest.tail()

subject                                                                                                                                  Government News  \
date                                               title                                                                                                   
Sep 9, 2017                                        DC Park Police Raid and Tear Down ‘Anti-Trump’ ...  Officials allegedly affiliated with the United...   
                                                   What Sign Language is This? Man Causes A Stir O...  The man giving out sign language during Govern...   
https://100percentfedup.com/12-yr-old-black-con... https://100percentfedup.com/12-yr-old-black-con...                                                NaN   
https://100percentfedup.com/video-hillary-asked... https://100percentfedup.com/video-hillary-asked...                                                NaN   
https://fedup.wpengine.com/wp-content/uploads/2... https://fedup.wpengine.com/wp-content/uploads/2...                                                NaN   

subject                                                                                                                                         politics  \
date                                               title                                                                                                   
Sep 9, 2017                                        DC Park Police Raid and Tear Down ‘Anti-Trump’ ...                                                NaN   
                                                   What Sign Language is This? Man Causes A Stir O...                                                NaN   
https://100percentfedup.com/12-yr-old-black-con... https://100percentfedup.com/12-yr-old-black-con...  https://100percentfedup.com/12-yr-old-black-co...   
https://100percentfedup.com/video-hillary-asked... https://100percentfedup.com/video-hillary-asked...  https://100percentfedup.com/video-hillary-aske...   
https://fedup.wpengine.com/wp-content/uploads/2... https://fedup.wpengine.com/wp-content/uploads/2...  https://fedup.wpengine.com/wp-content/uploads/...   

subject                                                                                               politicsNews  \
date                                               title                                                             
Sep 9, 2017                                        DC Park Police Raid and Tear Down ‘Anti-Trump’ ...          NaN   
                                                   What Sign Language is This? Man Causes A Stir O...          NaN   
https://100percentfedup.com/12-yr-old-black-con... https://100percentfedup.com/12-yr-old-black-con...          NaN   
https://100percentfedup.com/video-hillary-asked... https://100percentfedup.com/video-hillary-asked...          NaN   
https://fedup.wpengine.com/wp-content/uploads/2... https://fedup.wpengine.com/wp-content/uploads/2...          NaN   

subject                                                                                               worldnews  
date                                               title                                                         
Sep 9, 2017                                        DC Park Police Raid and Tear Down ‘Anti-Trump’ ...       NaN  
                                                   What Sign Language is This? Man Causes A Stir O...       NaN  
https://100percentfedup.com/12-yr-old-black-con... https://100percentfedup.com/12-yr-old-black-con...       NaN  
https://100percentfedup.com/video-hillary-asked... https://100percentfedup.com/video-hillary-asked...       NaN  
https://fedup.wpengine.com/wp-content/uploads/2... https://fedup.wpengine.com/wp-content/uploads/2...       NaN

In [19]:
ptest.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 9133 entries, ('Apr 1, 2015', 'MUSLIM WOMAN ARRESTED FOR SPITTING ON HER FELLOW WHITE WALMART SHOPPERS BUT WHAT HAPPENS NEXT IS MORE UPSETTING') to ('https://fedup.wpengine.com/wp-content/uploads/2015/04/entitled.jpg', 'https://fedup.wpengine.com/wp-content/uploads/2015/04/entitled.jpg')
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Government News  545 non-null    object
 1   politics         4382 non-null   object
 2   politicsNews     773 non-null    object
 3   worldnews        3590 non-null   object
dtypes: object(4)
memory usage: 400.0+ KB


In [20]:
ptest.shape

(9133, 4)

Using this, it is obvious that the news types do not need to be separated into their subject columns as some actually would be under multiple and it will not affect the overall results.  We will make a new ptest to remove this from the dataframe.


In [21]:
fer = test.drop('subject',axis=1)
fer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9427 entries, 0 to 9426
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   index   9427 non-null   int64 
 1   title   9427 non-null   object
 2   text    9427 non-null   object
 3   date    9427 non-null   object
dtypes: int64(1), object(3)
memory usage: 294.7+ KB


In [22]:
fern=pd.pivot_table(fer,index=['date'], aggfunc='first',values=['title','text'])
fern.tail()

text  \
date                                                                                                    
Sep 9, 2017                                         Officials allegedly affiliated with the United...   
https://100percentfedup.com/12-yr-old-black-con...  https://100percentfedup.com/12-yr-old-black-co...   
https://100percentfedup.com/video-hillary-asked...  https://100percentfedup.com/video-hillary-aske...   
https://fedup.wpengine.com/wp-content/uploads/2...  https://fedup.wpengine.com/wp-content/uploads/...   
https://fedup.wpengine.com/wp-content/uploads/2...  https://fedup.wpengine.com/wp-content/uploads/...   

                                                                                                title  
date                                                                                                   
Sep 9, 2017                                         DC Park Police Raid and Tear Down ‘Anti-Trump’...  
https://100percentfedup.com/12-yr-old-black-con...  https://100percentfedup.com/12-yr-old-black-co...  
https://100percentfedup.com/video-hillary-asked...  https://100percentfedup.com/video-hillary-aske...  
https://fedup.wpengine.com/wp-content/uploads/2...  https://fedup.wpengine.com/wp-content/uploads/...  
https://fedup.wpengine.com/wp-content/uploads/2...  https://fedup.wpengine.com/wp-content/uploads/...

In [23]:
def prepdf(z,t,r):
    
    z=pd.pivot_table(z,index=['date'], aggfunc='first',values=['title','text'])
    t=pd.pivot_table(t,index=['date'], aggfunc='first',values=['title','text'])
    r=pd.pivot_table(r,index=['date'], aggfunc='first',values=['title','text'])
    return [z,t,r]

In [24]:
[true,fake,test] = prepdf(true,fake,test)

In [25]:
test.head()

text  \
date                                                              
Apr 1, 2015   This woman s having trouble entering the Walma...   
Apr 1, 2016   AUDIO OF KARL ROVE ON HUGH HEWITT   Picking   ...   
Apr 1, 2017   https://www.youtube.com/watch?v=GPqQIlWksbgVer...   
Apr 10, 2015  Where was the media coverage when this black t...   
Apr 10, 2016  What could Hillary be saying that s so top sec...   

                                                          title  
date                                                             
Apr 1, 2015   MUSLIM WOMAN ARRESTED FOR SPITTING ON HER FELL...  
Apr 1, 2016   WHILE WE’RE DEBATING CRUZ VS TRUMP: Listen To ...  
Apr 1, 2017   BONKERS BERNIE SANDERS: Prioritizing Jobs Over...  
Apr 10, 2015  BREAKING: WHY DID MASSACHUSETTS OFFICIALS WAIT...  
Apr 10, 2016  HILLARY Uses A Noise Machine To Keep Us From H...

In [26]:
test.tail()

text  \
date                                                                                                    
Sep 9, 2017                                         Officials allegedly affiliated with the United...   
https://100percentfedup.com/12-yr-old-black-con...  https://100percentfedup.com/12-yr-old-black-co...   
https://100percentfedup.com/video-hillary-asked...  https://100percentfedup.com/video-hillary-aske...   
https://fedup.wpengine.com/wp-content/uploads/2...  https://fedup.wpengine.com/wp-content/uploads/...   
https://fedup.wpengine.com/wp-content/uploads/2...  https://fedup.wpengine.com/wp-content/uploads/...   

                                                                                                title  
date                                                                                                   
Sep 9, 2017                                         DC Park Police Raid and Tear Down ‘Anti-Trump’...  
https://100percentfedup.com/12-yr-old-black-con...  https://100percentfedup.com/12-yr-old-black-co...  
https://100percentfedup.com/video-hillary-asked...  https://100percentfedup.com/video-hillary-aske...  
https://fedup.wpengine.com/wp-content/uploads/2...  https://fedup.wpengine.com/wp-content/uploads/...  
https://fedup.wpengine.com/wp-content/uploads/2...  https://fedup.wpengine.com/wp-content/uploads/...

In [27]:
true.tail()

text  \
date                                                                    
September 7, 2017   WASHINGTON (Reuters) - President Donald Trump’...   
September 8, 2016   WASHINGTON (Reuters) - The White House on Thur...   
September 8, 2017   (Reuters) - A former top security official who...   
September 9, 2016   WASHINGTON/NEW YORK (Reuters) - Presidential c...   
September 9, 2017   WASHINGTON (Reuters) - President Donald Trump ...   

                                                                title  
date                                                                   
September 7, 2017   Trump's son met Russian lawyer for damaging in...  
September 8, 2016   White House names retired Air Force general as...  
September 8, 2017   California university system sues Trump over r...  
September 9, 2016   Clinton, Trump escalate fight in dramatic week...  
September 9, 2017   Trump says churches should get FEMA funds for ...

In [28]:
fake.tail(10)

text  \
date                                                                                                    
September 7, 2017                                   Donald Trump, Jr. went before Congress today t...   
September 8, 2016                                   Hillary Clinton is far from the perfect politi...   
September 8, 2017                                   Fox News has a problem with sexual predators i...   
September 9, 2016                                   Donald Trump is continuing his campaign of fea...   
September 9, 2017                                   Two Florida Republican lawmakers voted against...   
https://100percentfedup.com/12-yr-old-black-con...  https://100percentfedup.com/12-yr-old-black-co...   
https://100percentfedup.com/served-roy-moore-vi...  https://100percentfedup.com/served-roy-moore-v...   
https://100percentfedup.com/video-hillary-asked...  https://100percentfedup.com/video-hillary-aske...   
https://fedup.wpengine.com/wp-content/uploads/2...  https://fedup.wpengine.com/wp-content/uploads/...   
https://fedup.wpengine.com/wp-content/uploads/2...  https://fedup.wpengine.com/wp-content/uploads/...   

                                                                                                title  
date                                                                                                   
September 7, 2017                                    Don Trump Jr. Testified To The Senate Today A...  
September 8, 2016                                    In Brilliant Op-Ed, Hillary Clinton Opens Up ...  
September 8, 2017                                    Another Pervert Out At Fox News Over Allegati...  
September 9, 2016                                    Trump Gives Insane Warning On Religious Right...  
September 9, 2017                                    Two Florida Lawmakers Just Voted Against Hurr...  
https://100percentfedup.com/12-yr-old-black-con...  https://100percentfedup.com/12-yr-old-black-co...  
https://100percentfedup.com/served-roy-moore-vi...  https://100percentfedup.com/served-roy-moore-v...  
https://100percentfedup.com/video-hillary-asked...  https://100percentfedup.com/video-hillary-aske...  
https://fedup.wpengine.com/wp-content/uploads/2...  https://fedup.wpengine.com/wp-content/uploads/...  
https://fedup.wpengine.com/wp-content/uploads/2...  https://fedup.wpengine.com/wp-content/uploads/...

In [29]:
test = test.drop(test.tail(4).index)
test.tail()

text  \
date                                                             
Sep 8, 2016  These images were all taken from a Facebook pa...   
Sep 8, 2017  Hurricane Irma is moving a little west. The Ca...   
Sep 9, 2015  The reason the Baltimore Mayor gives for the h...   
Sep 9, 2016                                                      
Sep 9, 2017  Officials allegedly affiliated with the United...   

                                                         title  
date                                                            
Sep 8, 2016  SHOCKING: ENTIRE FACEBOOK PAGE DEDICATED To Ev...  
Sep 8, 2017  #Irma Update: Latest Look as Hurricane Strengt...  
Sep 9, 2015  (VIDEO) BALTIMORE MAYOR GIVES TOTALLY INSANE R...  
Sep 9, 2016  TREY GOWDY RIPS INTO THE DOJ AND FBI: “I was S...  
Sep 9, 2017  DC Park Police Raid and Tear Down ‘Anti-Trump’...

In [30]:
fake = fake.drop(fake.tail(5).index)
fake.tail()

text  \
date                                                                   
September 7, 2017  Donald Trump, Jr. went before Congress today t...   
September 8, 2016  Hillary Clinton is far from the perfect politi...   
September 8, 2017  Fox News has a problem with sexual predators i...   
September 9, 2016  Donald Trump is continuing his campaign of fea...   
September 9, 2017  Two Florida Republican lawmakers voted against...   

                                                               title  
date                                                                  
September 7, 2017   Don Trump Jr. Testified To The Senate Today A...  
September 8, 2016   In Brilliant Op-Ed, Hillary Clinton Opens Up ...  
September 8, 2017   Another Pervert Out At Fox News Over Allegati...  
September 9, 2016   Trump Gives Insane Warning On Religious Right...  
September 9, 2017   Two Florida Lawmakers Just Voted Against Hurr...

The excess information that does not include a date has been removed, these are websites that can be taken from, but will not be a part of this project for now.


In [32]:
test_uni = test.nunique()
test_uni

text     927
title    971
dtype: int64

In [33]:
drop = test.drop_duplicates(subset ="text", keep = False)

In [34]:
drop.nunique()

text     926
title    926
dtype: int64

In [35]:
test.nunique()

text     927
title    971
dtype: int64

In [36]:
test = drop

In [37]:
def get_uni(w,e):
    w = w.drop_duplicates(subset ="text", keep = False)
    e = e.drop_duplicates(subset ="text", keep = False)
    return [w,e]

In [38]:
[fake,true]=get_uni(fake,true)

In [40]:
print('Test  :')
print(test.nunique())
print(test.shape)

Test  :
text     926
title    926
dtype: int64
(926, 2)


In [41]:
print('Fake  :')
print(fake.nunique())
print(fake.shape)

Fake  :
text     1614
title    1614
dtype: int64
(1614, 2)


In [42]:
print('True :')
print(true.nunique())
print(true.shape)

True :
text     716
title    716
dtype: int64
(716, 2)


After reshaping, and getting rid of the duplicates, the values on this have went from over 40,000 to 3,256.  This is concerning and will be reviewed more later.

In [62]:
test.count()

text     926
title    926
dtype: int64

In [63]:
test.isna().sum()

text     0
title    0
dtype: int64

In [64]:
test.isnull().sum()

text     0
title    0
dtype: int64

In [65]:
(test == "").sum()

text     0
title    0
dtype: int64

In [68]:
def nulcnt(n,o):
    print(n.count(), n.isna().sum(),n.isnull().sum(),(n=="").sum())
    print(o.count(), o.isna().sum(),o.isnull().sum(),(o=="").sum())

In [69]:
nulcnt(fake,true)

text     1614
title    1614
dtype: int64 text     0
title    0
dtype: int64 text     0
title    0
dtype: int64 text     0
title    0
dtype: int64
text     716
title    716
dtype: int64 text     0
title    0
dtype: int64 text     0
title    0
dtype: int64 text     0
title    0
dtype: int64


Now we see that all of the information is unique and not null or otherwise empty.  the final will be placed into .csv files for later use.

In [70]:
test.to_csv('test.csv')
fake.to_csv('for_fake.csv')
true.to_csv('for_true.csv')